Fix Issue in Select-Object where Hashtable members (e.g., `Keys`) cannot be used with -Property or -ExpandProperty

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Prior to this PR, a hashtable piped into `Select-Object -[Expand]Property Keys` where the user expected `Keys` to retrieve the _property value_ from the hashtable rather than the _dictionary entry_ would return no value (in the case of -ExpandProperty, an error would be emitted, stating that the property could not be found).

This PR fixes this issue, and examines hashtable members as well as their key/value entries. The logic is designed such that hashtable key/values are still prioritised over hashtable members, so if a user creates a hashtable such as `@{ Keys = "None" }` and then selects the `Keys` property via Select-Object, it will have the same result as using `$ht.Keys` already does: the returned value will always prioritise an actual key/value entry over the hashtable members.

Note that this means creating a hashtable with a key named `Keys` will prevent access to the `Keys` member on the hashtable in this manner. This new behaviour is consistent with existing behaviour with retrieval of hashtable keys, and also consistent with dot-property access behaviour with hashtables -- you can access `$ht.Keys` up until you add a key with the name `Keys`, at which point you will always get that key value back.

Tests have been added to cover these cases explicitly.

## PR Context

Users should be able to utilise hashtables as actual hashtable objects as well as pseudo-PSObjects where appropriate. It is disingenuous for these commands to pretend hashtables don't have other members which may be desired in some circumstances.

This PR fixes #11094 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [x] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vexx32 committed:
:sparkles: Allow referencing hashtable members

- Fix an issue where Select-Object and similar could not resolve
member names from hashtables due to them carelessly converting them.

- Allows both hashtable keys/values and actual members to be referenced.
Precedence is given to the hashtable keys first to avoid collisions.

vexx32 committed:
:art: Whitespace fixes

Review this commit separately to the next.
Committed separately due to the file being old and needing a reformat.

vexx32 committed:
:white_check_mark: Add new tests

vexx32 committed:
:bug: Fix remaining bugs

- Fixed issue where -expandproperty found members but -property didn't

- Removed unnecessary second iteration of property names

vexx32 committed:
:white_check_mark: add missing CI tag

vexx32 commented:
Note for reviewers -- the Select-Object test file hasn't been updated in a _long_ time (still using tabs instead of spaces) so I included a format-only commit 3c1120c to sort that separately. You'll probably want to review that commit individually and then the following one for added tests.

No existing tests were modified in this PR apart from the formatting fixes. I only did a quick tabs-to-spaces and then auto-format commit, so if there are specific style issues you also want to fix while I'm there let me know, we can fix them either here or in a follow up PR as necessary. 🙂

vexx32 commented:
Also, I think the Codacy suggestion to make `IfHashtableWrapAsPSCustomObject()` a static method can be ignored. 😄

PoshChan commented on src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs:
```diff
@@ -159,27 +159,41 @@ public List<PSPropertyExpression> ResolveNames(PSObject target, bool expand)
 
```
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
Select-Object DRT basic functionality.Select-Object with empty script block property should throw
```powershell
Expected an exception, with FullyQualifiedErrorId 'EmptyScriptBlockAndNoName,Microsoft.PowerShell.Commands.SelectObjectCommand' to be thrown, but no exception was thrown.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1: line 131
131:             Should -Throw -ErrorId "EmptyScriptBlockAndNoName,Microsoft.PowerShell.Commands.SelectObjectCommand" -PassThru
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
Select-Object DRT basic functionality.Select-Object with empty script block property should throw
```powershell
Expected an exception, with FullyQualifiedErrorId 'EmptyScriptBlockAndNoName,Microsoft.PowerShell.Commands.SelectObjectCommand' to be thrown, but no exception was thrown.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Select-Object.Tests.ps1: line 131
131:             Should -Throw -ErrorId "EmptyScriptBlockAndNoName,Microsoft.PowerShell.Commands.SelectObjectCommand" -PassThru
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
Select-Object DRT basic functionality.Select-Object with empty script block property should throw
```powershell
Expected an exception, with FullyQualifiedErrorId 'EmptyScriptBlockAndNoName,Microsoft.PowerShell.Commands.SelectObjectCommand' to be thrown, but no exception was thrown.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1: line 131
131:             Should -Throw -ErrorId "EmptyScriptBlockAndNoName,Microsoft.PowerShell.Commands.SelectObjectCommand" -PassThru
```

vexx32 commented:
Hmm, interesting. Didn't expect to break that, but there you have it. 

@PoshChan remind me in 17 hours

vexx32 commented:
@PoshChan please remind me in 17 hours

iSazonov commented:
@vexx32 The PR description is more complex than the fix :-)

PoshChan commented:
@vexx32, this is the reminder you requested 17 hours ago

vexx32 committed:
:white_check_mark: Fix test that was broken

The reformat broke this test (it relies on being a truly empty scriptblock, no spaces at all).

Remove the added space to fix the test.

vexx32 commented:
That was simpler than I expected! 🎉

vexx32 commented:
@PoshChan please retry macos

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-macOS`

vexx32 committed:
:recycle: Address @daxian-dbw review comments

- Add test for Count & Length members
- Use base target object instead of wrapping hashtable

iSazonov commented:
@vexx32 Please look CI failures.

vexx32 committed:
:recycle: un-break test file

 - Still had tab stops in some places, replaced everything with spaces.
 - removed some accidentally duplicated code
 - 'Format Document' to ensure consistent indentation.

 Sorry about the diff, this file was a bit of a mess :pray:

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -362,13 +340,65 @@ Describe "Select-Object with Property = '*'" -Tags "CI" {
+        $hashtable | Select-Object -ExpandProperty Count | Should -Be 3
+    }
+
+    It 'should get the hashtable Count member' {
```
This one is a duplicate of the one above.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -362,13 +340,65 @@ Describe "Select-Object with Property = '*'" -Tags "CI" {
+    It 'should get the Length member from a hashtable' {
+        $hashtable = @{ a = 10; b = 20; c = 30 }
+
+        $result = $hashtable | Select-Object -Property Length
```
> Can you please add a test on selecting the Count and Length properties of a hashtable?

Sorry about the `Length` part in my comment, I thought there was a `Length` alias defined for `Count` but I was wrong.
You can remove this test.

vexx32 replied:
@daxian-dbw thanks for clarifying! I thought there was for a moment as well, but seems not.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -6,249 +6,227 @@ Add-TestDynamicType
-		$results[2].LastName | Should -Be $employees[2].LastName
-		$results[3].LastName | Should -Be $employees[3].LastName
+    It "Select-Object with empty script block property should throw" {
+        $e = { "bar" | Select-Object -Prop { } -ErrorAction Stop } |
```
This test failed in CI. It turns out `"bar" | Select-Object -Prop {} -ErrorAction Stop` throws, but `{ }` doesn't. It must be a bug :)

vexx32 replied:
Yeah, the check done in that particular case is a bit too simplistic, for sure. 😄

vexx32 replied:
The check is done here:

https://github.com/PowerShell/PowerShell/blob/f4b8ae5e048a5d2c2d8d06b36e4ac1c827346fdf/src/Microsoft.PowerShell.Commands.Utility/commands/utility/Select-Object.cs#L464-L495

It looks like it's mainly done this way to ensure properties with an empty string as the name can't be created. Technically a property name can be used if it's just a space, though it's not really discoverable... so I guess it might be intentional, though I question whether it's very useful functionality to enable that. 😂

daxian-dbw replied:
I'm fine keeping as is. Just update the test to use `{}` like before and that should be fine.

vexx32 committed:
:white_check_mark: Resolve test issues

daxian-dbw commented:
@vexx32 The test failure in mac CI is a known issue. Can you please check on the Windows failure?

vexx32 commented:
That's one I haven't seen before. Something to do with remoting? Not sure what to make of it 😅 

```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Management.Automation.Remoting.PrioritySendDataCollection.ReadOrRegisterCallback(OnDataAvailableCallback callback, DataPriorityType& priorityType) in D:\a\1\s\src\System.Management.Automation\engine\remoting\fanin\PriorityCollection.cs:line 240
   at System.Management.Automation.Remoting.Client.WSManClientCommandTransportManager.SendOneItem() in D:\a\1\s\src\System.Management.Automation\engine\remoting\fanin\WSManTransportManager.cs:line 3975

   at System.Management.Automation.Remoting.Client.WSManClientCommandTransportManager.OnRemoteCmdSendCompleted(IntPtr operationContext, Int32 flags, IntPtr error, IntPtr shellOperationHandle, IntPtr commandOperationHandle, IntPtr operationHandle, IntPtr data) in D:\a\1\s\src\System.Management.Automation\engine\remoting\fanin\WSManTransportManager.cs:line 3681
An error has occurred that was not properly handled. Additional information is shown below. The PowerShell process will exit.
Execution of {& $powershell $PSFlags -c $command} by build.psm1: line 1332 failed with exit code -1073741819
At D:\a\1\s\build.psm1:2191 char:17
+                 throw $errorMessage
+                 ~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : OperationStopped: (Execution of {& $po…it code -1073741819:String) [], RuntimeException
+ FullyQualifiedErrorId : Execution of {& $powershell $PSFlags -c $command} by build.psm1: line 1332 failed with exit code -1073741819
##[error]PowerShell exited with code '1'.
```

daxian-dbw commented:
@PoshChan please retry windows

PoshChan commented:
@daxian-dbw, did not find any matching pull request checks

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs:
```diff
@@ -159,27 +160,43 @@ public List<PSPropertyExpression> ResolveNames(PSObject target, bool expand)
+                {
+                    members = target.Members.Match(
+                        _stringValue,
+                        PSMemberTypes.Properties | PSMemberTypes.PropertySet | PSMemberTypes.Dynamic);
```
We evaluate this again. Maybe put this in a variable before line 172?

vexx32 replied:
This is evaluated against a different target object in the prior case.

iSazonov replied:
I mean `PSMemberTypes.Properties | PSMemberTypes.PropertySet | PSMemberTypes.Dynamic`

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/Utilities/Mshexpression.cs:
```diff
@@ -344,13 +354,16 @@ private PSPropertyExpressionResult GetValue(PSObject target, bool eatExceptions)
             {
-                target = (PSObject)(LanguagePrimitives.ConvertPSObjectToType(targetAsHash, typeof(PSObject), false, null, true));
+                wrapped = true;
+                return (PSObject)(LanguagePrimitives.ConvertPSObjectToType(targetAsHash, typeof(PSObject), false, null, true));
```
Please add names for const arguments.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -6,249 +6,227 @@ Add-TestDynamicType
-
-    It "Should be able to use the alias" {
-	{ $dirObject | select } | Should -Not -Throw
+        $result = $(Select-Object -InputObject $dirObject -Last $TestLength).Length
+        $expected = $dirObject.Length
+        { $dirObject | Select-Object } | Should -Not -Throw
```
`Should -Not -Throw` is not good test. 
Why do we remove one test below? And why do we need `$expected` above? Why do we change the tests at all?

daxian-dbw replied:
You need to keep the `It "Should be able to use the alias" {` here.

iSazonov reviewed (commented):
~~Tests contain many style fixes - please move it in separate PR.~~

daxian-dbw commented:
@iSazonov It was called out at https://github.com/PowerShell/PowerShell/pull/11097#issuecomment-554862128 that there is a format-only commit included.
I'm fine to keep that commit in this PR because this test file was seldom touched.

vexx32 committed:
:recycle: Address review comments

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -6,249 +6,227 @@ Add-TestDynamicType
-	$result   = $(Select-Object -InputObject $dirObject -Last $TestLength).Length
-	$expected = $dirObject.Length
-
-	$result | Should -Be $expected
```
`$result | Should -Be $expected` shouldn't be removed here.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -6,249 +6,227 @@ Add-TestDynamicType
+        $dirObject[$TestLength].Size | Should -Be ($orig2 + 1)
+    }
 
-	It "Select-Object with Property First Last Overlap should work"{
```
Why is this test case removed?

vexx32 replied:
Weird, not sure, will add it back.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -362,13 +340,48 @@ Describe "Select-Object with Property = '*'" -Tags "CI" {
         $obj.psobject.TypeNames[0] | Should -BeLike "Selected*"
         $obj.psobject.TypeNames[1] | Should -Not -BeLike "Selected*"
+        $p = Get-Process -Id $pid | Select-Object -Property Process* -ExcludeProperty ProcessorAffinity -ExpandProperty Modules
+        $p[0].psobject.Properties.Item("ProcessorAffinity") | Should -BeNullOrEmpty
```
The new code should be moved after `$obj.psobject.TypeNames[2] | Should ..` line below.

vexx32 replied:
Looks like this got duplicated from above (L#328-329) somehow. Sorry about that, really not sure what on earth happened here. Will remove.

daxian-dbw reviewed (changes requested):
I agree with @iSazonov, something is wrong in the changes of these couple of tests.

vexx32 committed:
:white_check_mark: Restore mangled tests

Some of the Select-Object tests got a bit mangled during format.
This commit should restore them all back to proper order.

vexx32 commented:
Apologies about the mangled tests, not sure how that happened. Should be all fixed up now. 💖 😊
